### PR TITLE
feat(psl): merge repeated namespace blocks

### PR DESCRIPTION
**Pull request by StevenMcClankerton**
Allow repeated same-name namespace blocks in a single PSL file to contribute declarations to one logical namespace.

## Changes
- Create each namespace once and extend its members; preserve duplicate-name errors across blocks and declaration kinds.
- Retain every authored namespace node/span in `NamespaceSymbol.declarations`; migrate SQL, Mongo, and Prisma 7 diagnostic consumers.
- Use null-prototype namespace/member dictionaries for safe identifier handling.
- Add focused linear tests for merging, duplicates, namespace isolation, and provenance.

## Why
Blocks organize declarations without creating separate scopes. Symbol collection owns merging; consumers retain access to every declaration location.

**Scope:** Single-file reopening only. No partial declarations, multi-file loading, changed target restrictions, serialized format changes, or README changes.

## Validation
- Final review fixes (11bc1cfe274): 50 parser symbol-table tests passed before and after; 14 existing SQL namespace tests passed after removal of the redundant cross-block test. Both package typechecks and lint scripts passed.
- Before the final review fixes: 2,231 affected tests passed across parser, SQL authoring, Mongo authoring, Prisma 7 importer, and language server.
- Independent verification: five package typechecks and 126 focused tests passed.
- Affected builds/lint, dependency lint and docs checks passed.
- Unfiltered package typecheck rollup encountered unresolved published imports in unchanged integration tests; package-only follow-up passed. Root example-inclusive typecheck previously blocked on a Prisma 7 linux-nixos engine download HTTP 404.
- Earlier full package suite had one CLI exit-code test fail; its isolated retry passed. Full workspace suite is not claimed green.

## Manual QA
Before the review simplification, real CLI checks confirmed identical JSON/declaration emission across consolidated/split/reversed layouts and correct duplicate-name and policy diagnostics. Current revision validated through affected suites and focused provenance tests; manual QA was not repeated after the metadata refactor.

## Summary by CodeRabbit

- **Bug Fixes**
  - Reopening a namespace now correctly combines its declarations and members, including declarations interleaved with other top-level definitions.
  - Duplicate members are consistently detected across all declarations, while the first declaration remains authoritative.
  - Names such as `__proto__` and `constructor` are handled correctly.
  - Unsupported or invalid namespace diagnostics are now reported separately for each declaration, with accurate source locations across supported database targets.
- **Tests**
  - Added coverage for namespace merging, duplicate handling, special names, and declaration-specific diagnostics.

### CODE REVIEW

File: packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts [hunk 1]
```diff
@@ -174,15 +175,29 @@ export function buildSymbolTable(options: BuildSymbolTableOptions): SymbolTableR
...
-        namespaces[name] = buildNamespace(
-          name,
-          declaration,
-          diagnostics,
-          sourceFile,
-          pslBlockDescriptors,
-        );
+        const taken = namespaceMemberNames.get(name) ?? new Set<string>();
+        namespaceMemberNames.set(name, taken);
+        Object.defineProperty(namespaces, name, {
```
Comment (SevInf) [inline packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts]: Why `defineProperty`?

File: packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts [hunk 2]
```diff
@@ -174,15 +175,29 @@ export function buildSymbolTable(options: BuildSymbolTableOptions): SymbolTableR
...
+          value: buildNamespace(
```
Comment (SevInf) [inline packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts]: why do we rebuild namespace from scratch every time isntead of extending it?

File: packages/1-framework/2-authoring/psl-parser/test/symbol-table.test.ts [hunk 3]
```diff
@@ -105,6 +105,240 @@ describe('buildSymbolTable() — AC3 namespace nesting', () => {
   });
 });
 
+describe('buildSymbolTable() — namespace reopening', () => {
```
Comment (SevInf) [inline packages/1-framework/2-authoring/psl-parser/test/symbol-table.test.ts:110]: You are way overusing `kind` and `it.each` in this test suite. If you have to do `if (kind == something)` within a test it probably should be a separate test.

Comment (SevInf) [inline packages/1-framework/2-authoring/psl-parser/test/symbol-table.test.ts:110]: I don't understand what any of those tests are testing.
We need to test that multiple blocks from different namespaces get merged and that block defined twice in the same namespace still produces the error. Test it as simple and as linear as possible, don't try to generalize the tests.

File: packages/1-framework/2-authoring/psl-parser/README.md [hunk 4]
```diff
@@ -56,6 +56,28 @@ Interpretation/validation (for example `@internal/sql-contract-psl`) is responsi
   - `@internal/psl-parser/syntax`
   - `@internal/psl-parser/tokenizer`
 
+## Reopening namespaces in one source file
```
Comment (SevInf) [inline packages/1-framework/2-authoring/psl-parser/README.md]: Stop inserting random-ass facts into readme. Read readme as a whole? Does fucking random detour about how namespaces are handled makes any sense in the place you are added? Will reader of this readme say, "perfect, I just wanted to know how namespaces are merged together right after you listed all public functions"

File: packages/2-sql/2-authoring/contract-psl/test/interpreter.block-attribute-requirements.test.ts [hunk 5]
```diff
@@ -66,6 +78,61 @@ function interpretWith(schema: string) {
 }
 
 describe('pslBlockDescriptors.requiresModelAttribute enforcement', () => {
+  const wrap = (body: string) => `namespace public {\n${body}\n}`;
+  const rule = 'audit_rule track_widgets {\n target = Widget\n}';
+  const model = (audited: boolean) =>
+    `model Widget {\n id Int @id\n @@map("widgets")\n${audited ? '@@audited' : ''}\n}`;
+
+  it.each([false, true])(
```
Comment (SevInf) [inline packages/2-sql/2-authoring/contract-psl/test/interpreter.block-attribute-requirements.test.ts]: Stop with `it.each`!
Why test this thing specifically? Why wouldn't it work over reopened namespace? We don't need to re-test every single bit of PSL logic just because we allow to have multiple PSL blocks now.

File: packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts [hunk 6]
```diff
@@ -140,6 +140,7 @@ export function buildSymbolTable(options: BuildSymbolTableOptions): SymbolTableR
   const models: Record<string, ModelSymbol> = {};
   const compositeTypes: Record<string, CompositeTypeSymbol> = {};
   const topLevelNames = new Set<string>();
+  const namespaceMemberNames = new Map<string, Set<string>>();
```
Comment (SevInf) [inline packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts]: We already know the names within namespace, i don't see a point of this map

File: packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts [hunk 7]
```diff
@@ -294,15 +310,17 @@ function buildNamespace(
...
-    models,
-    compositeTypes,
-    blocks,
-  };
+  return (
+    existing ?? {
+      kind: 'namespace',
+      name,
+      node,
+      span: nodePslSpan(node.syntax, sourceFile),
```
Comment (SevInf) [inline packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts]: Why does namespace has only one node and span? Namespace is declared in multiple places, we should be able to track all of them down.

File: packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts (hunk 1 above)
Comment (StevenMcClankerton) [reply packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts]: Replaced with ordinary assignment into a null-prototype namespace dictionary in fb966cc3c38.

File: packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts (hunk 2 above)
Comment (StevenMcClankerton) [reply packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts]: Simplified in fb966cc3c38: create the namespace once, then extend it in place with extendNamespace for each authored block.

File: packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts [hunk 8]
```diff
@@ -258,11 +273,12 @@ function buildNamespace(
...
+  existing: NamespaceSymbol | undefined,
+  taken: Set<string>,
 ): NamespaceSymbol {
-  const models: Record<string, ModelSymbol> = {};
-  const compositeTypes: Record<string, CompositeTypeSymbol> = {};
-  const blocks: Record<string, BlockSymbol> = {};
-  const taken = new Set<string>();
+  const models: Record<string, ModelSymbol> = existing?.models ?? {};
+  const compositeTypes: Record<string, CompositeTypeSymbol> = existing?.compositeTypes ?? {};
+  const blocks: Record<string, BlockSymbol> = existing?.blocks ?? {};
```
[the remark this answers is not in the document]
Comment (StevenMcClankerton) [reply packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts]: Fixed in fb966cc3c38: all namespace-member dictionaries have null prototypes. A focused regression checks enumerable own __proto__ members.

File: packages/1-framework/2-authoring/psl-parser/test/symbol-table.test.ts (hunk 3 above)
Comment (StevenMcClankerton) [reply packages/1-framework/2-authoring/psl-parser/test/symbol-table.test.ts:110]: Removed the added kind branches, it.each matrices, and projection helpers in fb966cc3c38. Replaced them with direct linear tests.

Comment (StevenMcClankerton) [reply packages/1-framework/2-authoring/psl-parser/test/symbol-table.test.ts:110]: Simplified in fb966cc3c38 to literal examples and direct assertions for merging, namespace isolation, duplicate errors, and declaration locations.

File: packages/1-framework/2-authoring/psl-parser/README.md (hunk 4 above)
Comment (StevenMcClankerton) [reply packages/1-framework/2-authoring/psl-parser/README.md]: Removed all README changes from this PR in fb966cc3c38, with no replacement documentation, per your follow-up.

File: packages/2-sql/2-authoring/contract-psl/test/interpreter.block-attribute-requirements.test.ts (hunk 5 above)
Comment (StevenMcClankerton) [reply packages/2-sql/2-authoring/contract-psl/test/interpreter.block-attribute-requirements.test.ts]: Removed these additions and restored the original test file and fixture in fb966cc3c38. Kept one SQL cross-block relation test and a diagnostic-location test for the new provenance shape.

File: packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts (hunk 6 above)
Comment (StevenMcClankerton) [reply packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts]: Removed the map in fb966cc3c38. Duplicate detection now checks the existing member dictionaries.

File: packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts (hunk 7 above)
Comment (StevenMcClankerton) [reply packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts]: Fixed in fb966cc3c38: NamespaceSymbol.declarations retains every authored node/span pair. Updated SQL, Mongo, and Prisma 7 diagnostic consumers to use those declaration locations, with focused provenance tests.

File: packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts [hunk 9]
```diff
@@ -252,22 +261,20 @@ function buildBlock(
...
-  const taken = new Set<string>();
+): void {
+  const { models, compositeTypes, blocks } = namespace;
+  namespace.declarations.push({ node, span: nodePslSpan(node.syntax, sourceFile) });
 
   for (const member of node.declarations()) {
     const memberName = member.name()?.name();
     if (memberName === undefined) continue;
-    if (taken.has(memberName)) {
+    if (memberName in models || memberName in compositeTypes || memberName in blocks) {
```
Comment (SevInf) [inline packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts]: Use `hasOwn`, not `in`

File: packages/2-sql/2-authoring/contract-psl/test/interpreter.namespaces.test.ts [hunk 10]
```diff
@@ -105,6 +105,45 @@ const baseInput = {
   capabilities: { sql: { scalarList: true } },
 } as const;
 
+describe('reopened namespace interpretation', () => {
```
Comment (SevInf) [inline packages/2-sql/2-authoring/contract-psl/test/interpreter.namespaces.test.ts]: again, why test this thing specifically?

File: packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts (hunk 9 above)
Comment (StevenMcClankerton) [reply packages/1-framework/2-authoring/psl-parser/src/symbol-table.ts]: Done in 11bc1cfe274b457b43589b898277e46a3b71e7a8: extendNamespace now uses Object.hasOwn for models, compositeTypes, and blocks. Existing parser tests pass (50); parser and SQL package typechecks/lints pass.

File: packages/2-sql/2-authoring/contract-psl/test/interpreter.namespaces.test.ts (hunk 10 above)
Comment (StevenMcClankerton) [reply packages/2-sql/2-authoring/contract-psl/test/interpreter.namespaces.test.ts]: Removed the entire reopened namespace interpretation describe block in 11bc1cfe274b457b43589b898277e46a3b71e7a8, without replacement. This file now matches its pre-PR content; all 14 remaining namespace tests pass.

**Review by SevInf — APPROVED** [review]
